Xlib/protocol/rq.py

```diff
@@ -600,8 +600,8 @@ def __init__(self, name, type, default = None):
     def parse_binary_value(self, data, display, length, format):
         return self.type.parse_binary(data, display)
 
-    def parse_value(self, val, display):
-        return self.type.parse_value(val, display)
+    def parse_value(self, val, display, **kwargs):
+        return self.type.parse_value(val, display, **kwargs)
```

Comment:
See commit description: https://github.com/python-xlib/python-xlib/pull/192/commits/417d85e2f82678c618ed3d583062a3369c5d3b6d

Reply:
OK, then it's better to add `rawdict=0` to all these methods. `**kwargs` should be used when number of arguments is really unpredictable/dynamic. Our case is pretty static.

Reply:
Done.

I took another look and it's a bit confusing that `Struct.parse_value` passes `rawdict` to the fields while `Struct.parse_binary` with a similar code doesn't. But I can't say that I understand the code and it's been working that way since 2007.

Reply:
I would suggest to make `parse_value` signature the same for all classes. Current fix may be dangerous for other extensions. At least some testing of all extensions would be useful. AFAIK current unit tests are very basic and mostly auto-generated. There are no unit tests for extensions, only examples.

Reply:
I just checked all instances of `rq.Object` in the current code and it's safe to assume that it's always wraps `rq.Struct`. Thus, the current fix will not result in passing `rawdict` to unrelated class.
Besides, we only get the issue here because of the nesting depth of XRes protocol objects. As far as I can see, there are no other instances of `Struct -> Object(Struct) -> Object(Struct)` in the code.
I can add `rawdict` to `Resource.parse_value` and `ResourceObj.parse_value`, but I won't have time to test the other extensions.

Another option is to comment `Xlib.ext.res.QueryResourceBytes`  and drop the `parse_value` patch. That would leave us with incomplete protocol support and 0 chance of regressions. I'm fine with that as I only ever needed `QueryClientIds` :smile: 

It's also interesting that the following patch is still passing all the tests:
```diff
@@ -1123,7 +1123,7 @@ class Struct(object):
                     field_val = val[vno:vno+f.structvalues]

                 if f.parse_value is not None:
-                    field_val = f.parse_value(field_val, display, rawdict=rawdict)
+                    field_val = f.parse_value(field_val, display)
                 ret[f.name] = field_val

             vno = vno + f.structvalues
```
I wonder if there's anyone who can remember why it's even trying to pass `rawdict`.

Reply:
I decided to try a different approach. Since the parser issue was caused by the object nesting level, I simply inlined one Struct and removed undesired nesting: https://github.com/python-xlib/python-xlib/pull/192/commits/80f90c971bc0965be5a83038facba1a0eefcdbc2

The change slightly deviates from the protocol spec but stays true to the wire format and carries all the same data. I believe it's fine to do that if we want to avoid risks. WDYT?

Reply:
Hmm, nice idea! Inline sub-Struct doesn't break the extension at functional level. I agree to go this way. Thanks!